linked_list/linked_list.py

```diff
@@ -2,99 +2,234 @@
 # Defines a node in the singly linked list
 class Node:
 
-    def __init__(self, value, next_node = None):
+    def __init__(self, value, next_node=None, previous=None):
         self.value = value
         self.next = next_node
+        self.previous = previous
 
 # Defines the singly linked list
+
+
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+        self.head = None  # keep the head private. Not accessible outside this class
+        self.tail = None
+        self.size = 0
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(n)
+
     def get_first(self):
-        pass
+        if self.head == None:
+            return None
 
+        return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(n)
+
     def add_first(self, value):

```

Comment:
👍 But the space complexity is O(1) since you're only ever increasing the size of the list by 1.